### ISSUE TITLE
Graph method for distances causes exception in Debug configuration

### ISSUE THREAD

**Issue by bwpge** [bwpge/aoc2023-cpp#2]
The [`Graph::distances`](https://github.com/bwpge/aoc2023-cpp/blob/d793396234386f22631615d2d772f382a7182266/day10/include/day10.hpp#L223) method causes a stack overflow exception in Debug configuration. In Release mode, this produces the correct output.

The method is likely reaching dangerous limits for recursion depth and should be fixed.

**Comment by bwpge** [comment]
An interesting side-effect of fixing this issue was that linked list destructors were also causing stack overflows. This is a known problem (see: https://softwareengineering.stackexchange.com/a/271239).

Fixed this issue by using a [vector of unique pointers](https://github.com/bwpge/aoc2023-cpp/blob/74d32362e4618973eb5f52729ae1f7657ba7f160/day10/include/day10.hpp#L273) to hold all nodes and only use raw pointers (e.g., non-owned) for adjacent/child nodes. This does pose a problem if we needed to remove/add nodes at random, but we don't. This is definitely a "quick fix" that works for the job and not much else.
